client/src/components/feedback/send.js

```diff
@@ -0,0 +1,220 @@
+    this.state = {
+      description: '',
+      isLoading: false,
+      timerCount: 3,
+      showTimer: true,
+      tooShortFeedback: false,
+      tooLongFeedback: false,
+      feedbackSent: false,
+      feedbackNotSent: false,
+      inputDisabled: false,
+      attemptCount: false,
+    };
+    this.onSubmit = this.onSubmit.bind(this);
+    this.handleChange = this.handleChange.bind(this);
+    this.tooShortFeedback = this.tooShortFeedback.bind(this);
+    this.tooLongFeedback = this.tooLongFeedback.bind(this);
+    this.sendApi = this.sendApi.bind(this);
+    this.cancel = this.cancel.bind(this);
+    this.showFeedbackSent = this.showFeedbackSent.bind(this);
+    this.showFeedbackNotSent = this.showFeedbackNotSent.bind(this);
+    this.showTimer = this.showTimer.bind(this);
+  }
+
+  onSubmit() {
+    if (this.state.description.length < 10) {
+      this.tooShortFeedback();
+    } else if (this.state.description.length > 400) {
+      this.tooLongFeedback();
+    } else {
+      this.sendApi();
+    }
+  }
+
+  handleChange(e) {
+    this.setState({
+      description: e.target.value,
+      tooShortFeedback: false,
+    });
+    // eslint-disable-next-line no-unused-expressions
+    e.target.value.length > 400
+      ? this.setState({
+        tooLongFeedback: true,
+      })
+      : this.setState({
+        tooLongFeedback: false,
+        attemptCount: false,
+      });
+  }
+
+  tooShortFeedback() {
+    const currentCallId = Math.random();
+    this.setState({
+      currentCallId,
+      tooShortFeedback: true,
+    });
+    setTimeout(() => {
+      if (currentCallId !== this.state.currentCallId) return;
+      this.setState({
+        tooShortFeedback: false,
+      });
+    }, 3000);
+  }
+
+  tooLongFeedback() {
+    const currentCallId = Math.random();
+    this.setState({
+      currentCallId,
+      attemptCount: true,
+      tooLongFeedback: true,
+    });
+    setTimeout(() => {
+      if (currentCallId !== this.state.currentCallId) return;
+      this.setState({
+        tooLongFeedback: true,
+        attemptCount: false,
+      });
+    }, 120);
+  }
+
+  sendApi() {
+    const currentCallId = Math.random();
+    this.setState({
+      isLoading: true,
+      tooShortFeedback: false,
+      tooLongFeedback: false,
+      showTimer: true,
+      inputDisabled: true,
+      currentCallId,
+    });
+    setTimeout(() => {
+      if (currentCallId !== this.state.currentCallId) return;
+      this.setState({
+        showTimer: false,
+      });
+    }, 4000);
+    setTimeout(() => {
+      if (currentCallId !== this.state.currentCallId) return;
+      Api.post('feedback', {
+        personId: 188,
+        restId: this.props.id,
+        description: this.state.description,
+      })
+        .then((response) => {
+          if (currentCallId !== this.state.currentCallId
+            && !this.state.isLoading
+            && response) return;
+          this.showFeedbackSent();
+        })
+        .catch((error) => {
```

Comment:
I am not sure. Lines 125-129 must be completed. That's why i situated them after 'possible' catch case.

Reply:
If you need to invoke some necessary code (close DB connection, close file, etc) you can use `.finally()`
https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise/finally

Reply:
https://habr.com/en/company/mailru/blog/269465/

Reply:
🤣 forgot about finally)
BTW, in some cases it is better to use try-with-recources instead of finally. You mentioned to close DB connection or fille - I mean this.